server/src/pages/shot/text-tool.js

```diff
@@ -0,0 +1,278 @@
+                                 this.state.textSelection.width,
+                                 this.state.textSelection.height);
+    this.drawingContext.fillStyle = `${FONT_COLOR}`;
+    this.drawingContext.font = `${FONT_WEIGHT} ${FONT_SIZE}px ${FONT_STYLE}`;
+    this.drawingContext.fillText(this.state.textContent, x, y);
+
+    if (this.props.confirmTextHandler) {
+      this.props.confirmTextHandler(this.state.textSelection, this.canvas.current, this.globalCompositeOperation || "source-over");
+    }
+    sendEvent("confirm-text", "text-toolbar");
+    this.reset();
+  }
+
+  drawText(e) {
+    const textSelection = new Selection(this.state.textSelection.left,
+                                        this.state.textSelection.top,
+                                        this.state.textSelection.left + e.target.clientWidth,
+                                        this.state.textSelection.top + e.target.clientHeight);
+    this.setState({textSelection});
+    this.setState({textContent: e.target.textContent});
+  }
+
+  onClickCancel(e) {
+    if (this.props.cancelTextHandler) {
+      this.props.cancelTextHandler();
+    }
+    sendEvent("cancel-text", "text-toolbar");
+    this.reset();
+  }
+
+  onChangeTextSize(event) {
+    const size = event.target.value;
+    this.setState({textSize: size});
+    this.props.textSizeCallback && this.props.textSizeCallback(size);
+    this.el.current.focus();
+  }
+
+  captureMousePosition(e) {
+    const boundingRect = this.canvas.current.getBoundingClientRect();
+    return {
+      x: e.clientX - boundingRect.left,
+      y: e.clientY - boundingRect.top
+    };
+  }
+
+  onMouseDown(e) {
+    if (e.button !== 0 || isMousedown) {
+      return;
+    }
+    isMousedown = true;
+    mousedownPosition = this.captureMousePosition(e);
+    e.preventDefault();
+
+    mousedownSelection = this.state.textSelection.clone();
+  }
+
+  onMouseUp(e) {
+    if (e.button !== 0) {
+      return;
+    }
+    if (e.target.id !== "text-input") {
+       e.target.style.cursor = "auto";
+    }
+    isMousedown = false;
+    mousedownPosition = null;
+    mousedownSelection = null;
+  }
+
+  onMouseMove(e) {
+    if (!isMousedown) {
+      return false;
+    }
+    e.target.style.cursor = "move";
+
+    const currentPosition = this.captureMousePosition(e);
+    const xDelta = currentPosition.x - mousedownPosition.x;
+    const yDelta = currentPosition.y - mousedownPosition.y;
+
+    const maxLeft = this.canvasCssWidth - mousedownSelection.width;
+    const maxTop = this.canvasCssHeight - mousedownSelection.height;
+    const newLeft = clamp(mousedownSelection.left + xDelta, 0, maxLeft);
+    const newTop = clamp(mousedownSelection.top + yDelta, 0, maxTop);
+
+    const updatedSelection = new Selection(
+      newLeft,
+      newTop,
+      newLeft + this.state.textSelection.width,
+      newTop + this.state.textSelection.height
+    );
+
+    this.setState({textSelection: updatedSelection});
+    return true;
+  }
```

Comment:
Is it possible to set `draggable` on `text-input` instead doing this?

Reply:
This worked for use case that's why didn't try Drag and Drop API, I can use and see if it helps to resolve some of the open issues